scripts/install.sh

```diff
@@ -0,0 +1,107 @@
+else
+  OS=linux
+fi
+
+# -------------------------------------------------------------------------------------------------
+# Helper to use the github redirect to find the latest release.
+function github_latest_release_tag() {
+  local PROJECT="$1"
+  curl \
+      -s \
+      -o /dev/null \
+      --write-out '%{redirect_url}' \
+      "https://github.com/${PROJECT}/releases/latest" \
+  | sed -e 's,https://.*/releases/tag/\(.*\),\1,'
+}
+
+# -------------------------------------------------------------------------------------------------
+# Helper to get a download url out of bazel build metadata file.
+function url_from_bazel_manifest() {
+  local MANIFEST_URL="$1"
+    if [[ "${OS}" == "darwin" ]]; then
+      local JSON_OS="macos"
+    else
+      local JSON_OS="ubuntu1404"
+    fi
+  wget -O - "${MANIFEST_URL}" \
+    | python -c "import json; import sys; print json.load(sys.stdin)['platforms']['${JSON_OS}']['url']"
+}
+
+# -------------------------------------------------------------------------------------------------
+# Helper to install bazel.
+function install_bazel() {
+  local VERSION="$1"
+
+  if [[ "${VERSION}" == "RELEASE" ]]; then
+    VERSION="$(github_latest_release_tag bazelbuild/bazel)"
+  fi
+
+  # macOS and trusty images have jdk8, so install bazel without jdk.
+  if [[ "${VERSION}" == "HEAD" ]]; then
+    # bazelbuild/continuous-integration/issues/234 - they don't seem to have an installed
+    # just raw binaries?
+    mkdir -p "${HOME}/bin"
+    wget -O "${HOME}/bin/bazel" \
+      "$(url_from_bazel_manifest https://storage.googleapis.com/bazel-builds/metadata/latest.json)"
+    chmod +x "${HOME}/bin/bazel"
+  else
+    wget -O install.sh \
+      "https://github.com/bazelbuild/bazel/releases/download/${VERSION}/bazel-${VERSION}-without-jdk-installer-${OS}-x86_64.sh"
+    chmod +x install.sh
+    ./install.sh --user
+    rm -f install.sh
+  fi
+
+  bazel version
+}
+
+# -------------------------------------------------------------------------------------------------
+# Helper to install buildifier.
+function install_buildifier() {
+  local VERSION="$1"
+
+  if [[ "${VERSION}" == "RELEASE" ]]; then
+    VERSION="$(github_latest_release_tag bazelbuild/buildtools)"
+  fi
+
+  if [[ "${VERSION}" == "HEAD" ]]; then
+    echo "buildifer head is not supported"
+    exit 1
+  fi
+
+  if [[ "${OS}" == "darwin" ]]; then
+    URL="https://github.com/bazelbuild/buildtools/releases/download/${VERSION}/buildifier.osx"
+  else
+    URL="https://github.com/bazelbuild/buildtools/releases/download/${VERSION}/buildifier"
+  fi
+
+  mkdir -p "${HOME}/bin"
+  wget -O "${HOME}/bin/buildifier" "${URL}"
+  chmod +x "${HOME}/bin/buildifier"
+  buildifier --version
+}
+
+# -------------------------------------------------------------------------------------------------
+# Install what is requested.
+[[ -z "${INSTALL_BAZEL:-}" ]] || install_bazel "${INSTALL_BAZEL}"
+[[ -z "${INSTALL_BUILDIFER:-}" ]] || install_buildifier "${INSTALL_BUILDIFER}"
```

Comment:
General nit:
most of the code seems to be prepared for invalid substitutions (eg. `${INSTALL_BAZEL:-}`), yet the script starts by setting `unset` options (`set -u`), which suggests the script should treat unset variables as if these were empty (which you use in line 19).

This creates a bit of oddity, because it has an effect on functions too. This means your functions would typically fail upon accessing `$1` (because of `set -e`), but you allow them to silently treat such errors as valid cases.

I would generally advise against `set -u` and encourage you to fix line 19 to check whether variable is set (or, better yet, use `uname` instead). This will work, but.. it enables pretty nasty behavior I'm not sure you want to use.

Reply:
Oh well, on second thought, I just realized `ksh` enables this by `set +u`, and disables by `set -u`. Ignore my previous comment :)